rfcs/taskforce/0001-buildkiteci.md

```diff
@@ -0,0 +1,215 @@
+4. Update branch protection on GitHub (only on develop)
+5. Building the wallet + bots (macos)
+6. Building the website (macos)
+7. Archive nodes
+8. Build daemon (macos)
+9. Build client-sdk
+10. Artifact build
+11. Unit tests
+12. Nonconsensus unit tests
+13. Various integration tests
+14. Nightly builds with larger configurations that we don't pay attention to
+
+The longer builds are unit tests (30min !!!) and integration tests (~20min average ;; 8min is just the build). The artifact medium curves build  (~45min !!!)
+
+### Breakdown of longest job (build-artifacts-medium-curves)
+
+1. Pinning opam packages ~2m
+2. Generate PV keys (build part1) ~5m30s
+3. Rebuild for PV keys changes (build part2) ~5m10s
+4. Generate runtime genesis ledger ~3m30s
+5. Build deb package ~2m30s
+6. Upload deb to repo ~1m30s
+7. Docker build and publish ~20m (!!!)
+
+## Details
+
+Mac machines are a pain to manage ourselves. For our mac jobs, keep using CircleCI.
+
+For everything else, switch to [BuildKite](https://buildkite.com)
+
+### Why Buildkite
+
+- Very good UI/UX around builds
+- Easy-to-install Agent model, so very easy to run agents on variety of hardware including cloud, local machines (for testing purposes) — includes first-party Helm charts for trivial deployment with our infrastructure
+- First-class artifact upload/download with GCP cloud storage
+- Dynamic pipelines allow dynamic reconfiguration of builds from builds themselves — this can be used for expanding work based on files touched and support generated YAML
+- Healthy ecosystem of plugins [https://buildkite.com/plugins](https://buildkite.com/plugins)
+- Supports direct encoding of dependency graph of jobs
+- Run buildkite agents locally or on manually spun up compute instances to test rules by targeting specific agents with tags and using the CLI [https://github.com/buildkite/cli](https://github.com/buildkite/cli)
+- Brandon has extensive experience working with BuildKite at Pinterest (and with a few side projects); lower risk
+
+## Details
+
+### Deploying of Nodes
+
+Deploy agents with a helm chart on k8s — same as the rest of our infrastructure [https://github.com/buildkite/charts](https://github.com/buildkite/charts). It sets a service account so we can still build docker images and use kubectl and helm during pipelines.
+
+Engineers who are iterating on build rules can run nodes locally or spin up instances to test things using the buildkite CLI
+
+### Description of pipelines
+
+Pipeline are configured with Dhall rather than jinja-templated yaml. Dynamic pipelines support this as a first-class citizen, we do not require running the equivalent "`[test.py](http://test.py)`" before pushing. I have already implemented this on a [side project](https://github.com/bkase/gameboy/blob/master/.buildkite/pipeline.yaml).
+
+**General Pipeline Writing Things**
+
+* Use the [BuildKite monorepo plugin](https://github.com/chronotc/monorepo-diff-buildkite-plugin) for granular test execution.
```

Comment:
This plugin automatically splits up builds by folders of the repo? That seems very expensive given the number of folders we have in `src/lib/`. I think we should consider grouping libs together better before enabling this.

Reply:
I was thinking we could split at a macro level at first `src/`, `frontend/` etc.

Future work to get it more fine-grained

Reply:
Ok. Can you control that with that plugin? That plugin description seemed to imply it does all folders, but maybe that's just the default behavior?

Reply:
My understanding is that the plugin lets you schedule new work that match one or more paths exposed via some diffing mechanism. I don't think it is ever iterating over all the folders in any way that we need to be concerned with. Does this address your question or am I missing something?

Reply:
I think I just misread the config examples. The tagline of the repo is "Run separate pipelines for each folder in your monorepo", which tripped me up when I started reading the examples.